ksqldb-functional-tests/src/test/resources/query-validation-tests/upgrades.json

```diff
@@ -201,9 +214,79 @@
         "CREATE TABLE bar AS SELECT * FROM foo;",
         "CREATE OR REPLACE TABLE bar AS SELECT * FROM foo WHERE col1 = '123';"
       ],
+      "inputs": [],
+      "outputs": []
+    },
+    {
+      "name": "DML - TABLE - add column",
+      "properties": {
+        "ksql.create.or.replace.enabled": true
+      },
+      "statements": [
+        "CREATE TABLE foo (id VARCHAR PRIMARY KEY, col1 VARCHAR, col2 VARCHAR) WITH (kafka_topic='foo', value_format='JSON');",
+        "CREATE TABLE bar AS SELECT id, col1 FROM foo;",
+        "CREATE OR REPLACE TABLE bar AS SELECT id, col1, col2 FROM foo;"
+      ],
+      "inputs": [],
+      "outputs": []
+    },
+    {
+      "name": "DML - TABLE - add column via select *",
+      "properties": {
+        "ksql.create.or.replace.enabled": true
+      },
+      "statements": [
+        "CREATE TABLE foo (id VARCHAR PRIMARY KEY, col1 VARCHAR, col2 VARCHAR) WITH (kafka_topic='foo', value_format='JSON');",
+        "CREATE TABLE bar AS SELECT id, col1 FROM foo;",
+        "CREATE OR REPLACE TABLE bar AS SELECT * FROM foo;"
+      ],
+      "inputs": [],
+      "outputs": []
+    },
+    {
+      "name": "DML - TABLE - remove column",
+      "properties": {
+        "ksql.create.or.replace.enabled": true
+      },
+      "statements": [
+        "CREATE TABLE foo (id VARCHAR PRIMARY KEY, col1 VARCHAR, col2 VARCHAR) WITH (kafka_topic='foo', value_format='JSON');",
+        "CREATE TABLE bar AS SELECT id, col1, col2 FROM foo;",
+        "CREATE OR REPLACE TABLE bar AS SELECT id, col1 FROM foo;"
+      ],
       "expectedException": {
         "type": "io.confluent.ksql.util.KsqlException",
-        "message": "Upgrades not yet supported for io.confluent.ksql.execution.plan.TableSink"
+        "message": "Cannot REPLACE data source: DataSource '`BAR`' has schema = `ID` STRING KEY, `COL1` STRING, `COL2` STRING which is not upgradeable to `ID` STRING KEY, `COL1` STRING(The following columns are changed or missing: [`COL2` STRING])"
+      }
+    },
+    {
+      "name": "DML - TABLE - change source",
+      "properties": {
+        "ksql.create.or.replace.enabled": true
+      },
+      "statements": [
+        "CREATE TABLE foo (id VARCHAR PRIMARY KEY, col1 VARCHAR) WITH (kafka_topic='foo', value_format='JSON');",
+        "CREATE TABLE baz (id VARCHAR PRIMARY KEY, col1 VARCHAR) WITH (kafka_topic='baz', value_format='JSON');",
+        "CREATE TABLE bar AS SELECT id, col1 FROM foo;",
+        "CREATE OR REPLACE TABLE bar AS SELECT id, col1 FROM baz;"
+      ],
+      "expectedException": {
+        "type": "io.confluent.ksql.util.KsqlException",
+        "message": "Query is not upgradeable. Plan step of type class io.confluent.ksql.execution.plan.TableSource must have matching topicName. Values differ: foo vs. baz"
+      }
+    },
+    {
+      "name": "DML - TABLE - change timestampColumn",
```

Comment:
Can we add a negative test for:
*  the timestamp format changing?
* bigger changes, like adding:
  * Join
  * Partition by
  * GROUP BY
  * etc.

Reply:
I added a few, but the implementation is an allowlist, not a blacklist, so it's pretty clear that anything we don't implement will throw